plugins/git-plugin/src/main/groovy/org/rundeck/plugin/scm/git/ssh/SshjSessionFactory.groovy

```diff
@@ -0,0 +1,92 @@
+package org.rundeck.plugin.scm.git.ssh
+
+import groovy.transform.CompileStatic
+import net.schmizz.keepalive.KeepAliveProvider
+import net.schmizz.sshj.DefaultConfig
+import net.schmizz.sshj.SSHClient
+import net.schmizz.sshj.transport.verification.PromiscuousVerifier
+import net.schmizz.sshj.userauth.keyprovider.KeyProvider
+import org.eclipse.jgit.errors.TransportException
+import org.eclipse.jgit.transport.CredentialsProvider
+import org.eclipse.jgit.transport.OpenSshConfig
+import org.eclipse.jgit.transport.RemoteSession
+import org.eclipse.jgit.transport.SshSessionFactory
+import org.eclipse.jgit.transport.URIish
+import org.eclipse.jgit.util.FS
+
+
+@CompileStatic
+class SshjSessionFactory extends SshSessionFactory {
+    private byte[] privateKey
+    Map<String, String> sshConfig
+    private OpenSshConfig config
+    private String privateKeyFile
+    private SSHClient sshClient
+
+    SshjSessionFactory(byte[] privateKey, Map<String, String> sshConfig) {
+        this.privateKey = privateKey
+        this.sshConfig = sshConfig
+
+        if(privateKey){
+            File tempFile = File.createTempFile("tmp", "key")
+            tempFile.deleteOnExit()
+            FileOutputStream fos = new FileOutputStream(tempFile)
+            fos.write(privateKey)
+            privateKeyFile = tempFile.getAbsolutePath()
+        }
+    }
+
+    @Override
+    RemoteSession getSession(URIish uri, CredentialsProvider credentialsProvider, FS fs, int tms) throws TransportException {
+
+        String user = uri.getUser()
+        String host = uri.getHost()
+        int port = uri.getPort()
+
+        if (config == null)
+            config = OpenSshConfig.get(fs)
+
+        final OpenSshConfig.Host hc = config.lookup(host)
+        if (port <= 0)
+            port = hc.getPort()
+        if (user == null)
+            user = hc.getUser()
+
+        sshClient = createConnection(user, host,port)
```

Comment:
Is there any specific reason to make `sshClient` a property of the factory instance? The property value is overwritten every time `getSession` gets called, which means the old one got discarded.

A session factory is better to be an immutable singleton object without saving a SSHClient as its internal state.

According to the document about SSHClient and RemoteSession, they seem to be an abstraction of the same thing - an established SSH tunnel connected to a host. I think it's natural that an SSHClient instance should be part of the SshjSession instance.  

```suggestion
        return new SshjSession(createConnection(user, host, port)
```